File: Day 14/Part_Two.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long int
const int MOD =  1e9 + 7;
const int N = 1e5 + 5;
const int INF = 1e10;
  
void compute()
{

}

int calc_load(vector <string> &arr)
{
    int n = arr.size();
    int ans = 0;

    for(int i=0; i<arr[0].size(); i++)
    {
        for(int j=0; j<n; j++)
        {
            if(arr[j][i] == 'O')
            {
                ans += (n - j);
            }
        }
    }

    return ans;
}

void cycle(vector <string> &arr)
{
    int n = arr.size();
    int m = arr[0].size();

    // North
    for(int temp=0; temp<=100; temp++)
    {    
        for(int i=0; i<m; i++)
        {
            for(int j=0; j<n-1; j++)
            {
                if(arr[j][i] == '.' && arr[j+1][i] == 'O') 
                {
                    arr[j][i] = 'O';
                    arr[j+1][i] = '.';
                }
            }
        }
    }

    // West
    for(int temp=0; temp<=100; temp++)
    {
        for(int i=0; i<n; i++)
        {
            for(int j=0; j<m-1; j++)
            {
                if(arr[i][j] == '.' && arr[i][j+1] == 'O') 
                {
                    arr[i][j] = 'O';
                    arr[i][j+1] = '.';
                }
            }
        }
    }

    // South
    for(int temp=0; temp<=100; temp++)
    {
        for(int i=0; i<m; i++)
        {
            for(int j=n-1; j>0; j--)
            {
                if(arr[j][i] == '.' && arr[j-1][i] == 'O') 
                {
                    arr[j][i] = 'O';
                    arr[j-1][i] = '.';
                }
            }
        }
    }

    // East
    for(int temp=0; temp<=100; temp++)
    {
        for(int i=0; i<n; i++)
        {
            for(int j=m-1; j>0; j--)
            {
                if(arr[i][j] == '.' && arr[i][j-1] == 'O') 
                {
                    arr[i][j] = 'O';
                    arr[i][j-1] = '.';
                }
            }
        }
    }
}

void solve()
{
    int n = 100;
    
    vector <string> arr(n);
    for(int i=0; i<n; i++) cin >> arr[i];

    int ans = 0;
    
    unordered_map <string, int> mp, freq;

    // This part was inspired from https://www.youtube.com/watch?v=4Ms3SN0lpxE [Jonathan Paulson]
    // Tried memorization

    for(int i=1; i<=1000000000; i++)
    {
        cycle(arr);
        ans = calc_load(arr);

        string temp;
        for(auto i : arr) temp += i;

        if(mp.find(temp) != mp.end() && freq[temp] >= 5)
        {
            int diff = i - mp[temp];
            
            int next = ((1000000000 - i) / diff)  * diff;
            i += next; 
        }

        freq[temp]++;
        mp[temp] = i;
    }

    cout << ans << '\n';
}
 
signed main(void) 
{
#ifndef ONLINE_JUDGE
    freopen("input.txt","r",stdin);
    freopen("output.txt","w",stdout);
#endif
 
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int t = 1;
    // cin >> t;

    compute();
    
    for(int i=0;i<t;i++)
    {
        // cout << "Case #" << i+1 << ": "; // Kick Start
        solve();
    }

	return 0;
}

File: Day 19/Part_Two.py
import sys
import math
sys.stdin = open("input.txt", "r")
sys.stdout = open("output.txt", "w")
sys.stderr = open("output.txt", "w")

with open("input.txt", "r") as f:
    data = f.read().split("\n\n")
    
    compute = data[0].split()
    query = data[1].split()

mp = {}
for i in compute:
    mp[i.split("{")[0]] = i.split("{")[1][:-1].split(',')

ans = 0
start = "in"
ranges = {"x":[1, 4000], "m":[1, 4000], "a":[1, 4000], "s":[1, 4000]}

def dfs(start, ranges):    
    global ans

    if(start == 'A'):
        ans += math.prod([ranges[j][1] - ranges[j][0] + 1 for j in ranges])
        return

    if(start == 'R'):
        return

    l = mp[start]
    for j in l:
        if(j.isalpha()):
            dfs(j, ranges)
        else:
            temp = j.split(":")[0]
            next = j.split(":")[1]

            if(temp[1] == "<"):
                c = ranges.copy()
                c[temp[0]] = [c[temp[0]][0], int(temp[2:])-1]
                ranges[temp[0]] = [int(temp[2:]), ranges[temp[0]][1]]
                dfs(next, c)
            elif(temp[1] == ">"):
                c = ranges.copy()
                c[temp[0]] = [int(temp[2:])+1, c[temp[0]][1]] 
                ranges[temp[0]] = [ranges[temp[0]][0], int(temp[2:])]
                dfs(next, c)

dfs(start, ranges)
print(ans)
